### ISSUE TITLE
ConfigArguments singleton not compatible with multiprocessing_context of spawn or forkserver

### ISSUE THREAD

**Issue by krehm** [argonne-lcf/dlio_benchmark#116]
I have tried running dlio_benchmark with infiniband, and need to use multiprocessing_context=spawn or forkserver.  It turns out that the ConfigArguments singleton doesn't play well with these modes.

When dlio_benchmark starts up, main instantiates the singleton, which then contains all the configuration default values.   main then calls LoadConfig() with the yaml file as a parameter, which causes attributes in the ConfigArguments singleton to be replaced with the overriding values from the yaml file and from the command line.

This all works fine in the main process, but when a child process is spawned, the pickled data sent to that child for execution only contains a reference to ConfigArguments, not the current contents of the parent's ConfigArguments singleton.  

I set a breakpoint at the beginning of method worker_init() in torch_data_loader.py.  When the child process breaks in pdb, I step into the ReaderFactory.get_reader() function, then step into the following line of code:

_args = ConfigArguments.get_instance()

Since this is a child process that was not created with fork, the ConfigArguments singleton doesn't yet exist, so a new one is created, but this new instantiation contains only the default config parameters, all of the overrides in the yaml file and on the command line are lost.  So the child process ends up executing incorrect code.

Any ideas on how to fix this?

Thanks, Kevan

**Comment by hariharan-devarajan** [comment]
@krehm For that the worker init code would have to load up the configuration and run `LoadConfig(self.args, cfg)` from the config module. here the cfg is the hydra configuration that needs to be loaded by each worker.
We can add member variable within ConfigurationManager to hold this config and then pass it to the worker_init function.
This should help u have a stateless worker (with spawn or fork server) and load the correct configuration on initialization of the worker.

**Comment by krehm** [comment]
@hariharan-devarajan if you could describe in a bit more detail how one would pickle the singleton and send it inside the working init, I would be willing to try to code it up.

**Comment by hariharan-devarajan** [comment]
So can u write a new data_loader implementation that uses the spawn.
We have a couple of options which need to be evaluated using DLIO-profiler.

1. Pass config directly to worker. (This is simplest to do but probably non-performant)
2. Make config idempotent, we need to identify random number behavior before hand and then we can just reply on worker in it. (My preferred)
3. Make config pickled, this will store a state down in a temporary file and then recreate it in the spawned object.

@krehm @zhenghh04 Thoughts?

**Comment by krehm** [comment]
First I will need to familiarize myself with the code better, then I may have a few questions.  Might be a few days, holiday schedules are kicking in.

**Comment by krehm** [comment]
@hariharan-devarajan, I have a related question.  I've been trying to figure out a workaround for this problem so that in the meantime we can make some benchmark runs.  unet3d normally runs with read_threads=4, so it occurred to me that if I set read_threads=0 but then run 4 times as many MPI processes, that I should end up with the equivalent number of sample readers running in parallel during the benchmark.  I guessed that that should provide similar performance and should exhibit similar results.  However, when I make such a run, I get:
```
[INFO] Averaged metric over all epochs
[METRIC] ==========================================================
[METRIC] Training Accelerator Utilization [AU] (%): 50.6920 (0.1175)
[METRIC] Training Throughput (samples/second): 59.5674 (0.1139)
[METRIC] Training I/O Throughput (MB/second): 8328.0802 (15.9269)
**[METRIC] train_au_meet_expectation: fail**
[METRIC] ==========================================================
```
Is there something wrong with my logic, this configuration is not really equivalent to the default unet3d configuration, or perhaps might it point to some issue with the code?

**Comment by zhenghh04** [comment]
@krehm below is the explanation. 

read_threads means the number of threads per MPI process (i.e., number of threads per GPU). 

If you set read_threads = 4, then 4 threads are loading data for one GPU. Increasing MPI processes increases the need for data proportionally. 

If you set read_threads = 0, the master threads will read data for one GPU, and there is no overlap between I/O and compute. That's why you are getting about 50% AU. 

For the throughput, it can be calculated as AU*4 (samples per batch)/1.36 (second per batch) * nproc.

**Comment by krehm** [comment]
Thanks much, that makes sense.

**Comment by hariharan-devarajan** [comment]
@krehm Try and run with read_threads: 1 This should give u a similar effect with 4 processes each having a master thread for compute and a child worker for I/O. So it will have 4 compute threads and 4 I/O threads in total.

Makes sense?

**Comment by krehm** [comment]
The underlying problem for us is that we can't currently run any torch benchmarks with a value other than read_threads=0.    The Infiniband libfabric library uses pthreads, so if we run with a non-zero read_threads value  in combination with 'fork',  then the benchmark hangs or throws backtraces.   And we can't run with 'spawn' or 'forkserver' either because of this issue.  read_threads=0 is our only choice.

**Comment by hariharan-devarajan** [comment]
I am just curious about this issue. Would this happen even with TensorFlow? What if u switched the framework and data loader to TensorFlow? I know the behavior of frameworks are different but it would work. 

Meanwhile, do u have any example of using `multiprocessing_context=spawn` or `forkserver`. I will be happy to see if i can have a quick solution to make it work with that.

**Comment by krehm** [comment]
Yes, tensorflow works fine, it doesn't fork when read_threads is > 0, it spins pthreads instead, so no issue.   We would like to us pytorch with unet3d so we can compare our benchmarks against others (MLPerf).

Just do a unet3d run but set multiprocessing_context to 'spawn'.   unet3d is pytorch, but because the child processes end up with a default self.args, the child process tries to  use tensorflow.

**Comment by hariharan-devarajan** [comment]
Where do u set that ? On the data loader do u have a link to Pytorch doc I will try to look at it

**Comment by krehm** [comment]
multiprocessing_context is set under the 'reader' section.   Some links:
[this](https://pytorch.org/docs/stable/notes/multiprocessing.html) describes how you can't use fork with CUDA.   [This](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) covers multiprocessing_context in detail.

**Comment by krehm** [comment]
and [this](https://discuss.python.org/t/switching-default-multiprocessing-context-to-spawn-on-posix-as-well/21868) talks about the fundamental problem of using the 'fork' method.

**Comment by hariharan-devarajan** [comment]
So the idea is to use multi procession on the data loader and then use r threads per multi process?

If we have a train method which builds data loader and runs training loop we have to add a multi processing on top of it

```
def train():
  # build data loader
  # train loop

Def main():
  Multi process on train 
```

Is the idea correct. Would this make sense to have it for any framework like tensorflow or for dali? Is it specific to Pytorch

**Comment by hariharan-devarajan** [comment]
If this is correct then, what is the expected behavior for data parallelism? Cause the data loader does the distribution. In this case we have to manually distribute among spawned processes and then do again in data loader. 
Also in this case if you run with read threads =0 it will end up with same behavior as mpirun -n 4 with reads threads 0.

Am I missing something?


This feels very weird overall.

**Comment by hariharan-devarajan** [comment]
What I initially thought you wanted was to set

`torch.multiprocessing.set_start_method('spawn')` 

So that the data loader uses spawn instead of fork

**Comment by krehm** [comment]
In function read() in torch_data_loader.py is where the value of 'multiprocessing_context' gets passed to the torch DataLoader.  The torch DataLoader then starts child processes using one of fork, spawn, or forkserver, and it can optionally pass arguments to each of the started processes's init function.  The torch code appears to support all three modes.  That's about all I know so far, and yes, somehow pass the configuration to the child init function in a way that preserves all the values that were added via LoadConfig().   Still learning this stuff as I go.

**Comment by krehm** [comment]
@hariharan-devarajan see pull request 129 for a fix for the spawn/forkserver problem, see if it is acceptable.

**Comment by zhenghh04** [comment]
This is fixed by pull#129. I am closing this one now.
